zingo-sync/src/scan/compact_blocks/runners.rs

```diff
@@ -96,42 +104,42 @@ where
         for tx in block.vtx.into_iter() {
```

Comment:
sorry that this might get more complicated but this add_block method seems specific to trial decryption. in ShardTreeData which is returned in ScanResults of a scan, we have the initial leaf position and a vec of Nodes and Retentions. I think we want to massage the way that batchrunners add widgets to allow to add all outputs from a compact block for trial decryption but for shard tree batchs we want to add an initial position for the Batch and a bunch of Nodes and Retentions. maybe the simple way to do this would be to move the pre-processing to outside of the batchrunner completely so for example in the trial decrpytion case we gather all the outputs (batch inputs) and feed directly to batch runners? this way the batch runners doesnt need additional traits for pre-processing batch inputs

Reply:
i think its worth also looking into the best format for inserting into ShardStore in a parallel manner, as initial positon, Nodes and Retention is specific for `batch_insert` and it might be better to include the position with each node/retention for example

Reply:
> i think its worth also looking into the best format for inserting into ShardStore in a parallel manner, as initial positon, Nodes and Retention is specific for `batch_insert` and it might be better to include the position with each node/retention for example

The best method of parallelizing I believe is to create `LocalizedPrunableTree`s, which can be created in parallel (each one taking an initial position (well, a range, but the upper bound is unnecessary), as well as Nodes and Retentions), and then acquiring the shardtree lock at the end in order to insert the localizedprunabletrees. The trees will have been constructed and maximally pruned in parallel, and then insertion is still unparallelizable, but after pruning there will be probably at least an order of magnitude fewer nodes to insert
.

Reply:
> sorry that this might get more complicated but this add_block method seems specific to trial decryption. in ShardTreeData which is returned in ScanResults of a scan, we have the initial leaf position and a vec of Nodes and Retentions. I think we want to massage the way that batchrunners add widgets to allow to add all outputs from a compact block for trial decryption but for shard tree batchs we want to add an initial position for the Batch and a bunch of Nodes and Retentions. maybe the simple way to do this would be to move the pre-processing to outside of the batchrunner completely so for example in the trial decrpytion case we gather all the outputs (batch inputs) and feed directly to batch runners? this way the batch runners doesnt need additional traits for pre-processing batch inputs

Well, we could do a similar thing where we add the batch inputs as we acquire them. It might be simpler in the concrete case to build up a full batch's worth of nodes instead of having the add_widgets(name pending) method, but I think using the same interface for both will be simpler than trying to make the interfaces diverge further.